### ISSUE TITLE
Add support for numeric character references to `twisted.web.template`

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5408]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5408|
|Type|enhancement|
|Created|2011-11-23 15:54:03Z|
|Branch|https://github.com/twisted/twisted/tree/numcharref-5408|

To let applications emit numeric character references (`&#1234;`), `twisted.web.template` needs an expanded interface.  A string containing `&#1234;` will be flattened to `&amp;#1234;` (correctly).  There needs to be a way to express to the flattener that a numeric character reference is desired.

The API should be pretty simple.  eg,

```
nbsp = twisted.web.template.entity(ord(u'\N{NO-BREAK SPACE}'))
```

or perhaps even just

```
nbsp = twisted.web.template.entity(u'\N{NO-BREAK SPACE}')
```

Flattening such an object should produce a numeric character reference, eg `&#160;`.

<details><summary>Searchable metadata</summary>

```
trac-id__5408 5408
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_numcharref_5408 branches/numcharref-5408
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__web web
keywords__easy easy
time__1322063643000000 1322063643000000
changetime__1325958929000000 1325958929000000
version__None None
owner__exarkun exarkun
cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="amberite's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|amberite commented|
|-|-|

Calling it twisted.web.template.entity is perhaps imprecise, in that the markup world refers to them as "character entity reference" and "numeric character reference", both of which are lumped under the term "character reference", so maybe charref is the thing to call it.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

A clarification regarding the use-case: the goal is to be able to flatten to ASCII, even in the presence of non-ASCII characters.  Explicit numeric character references are one approach, and maybe supporting them explicitly is a good idea regardless.  Another approach which occurs to me, though, is that the flattener could encode using `ascii` with the `xmlcharrefreplace` error handler.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

(In [33389]) Branching to 'numcharref-5408'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

(In [33391]) Push `__all__` to the top of the file, so perhaps people will remember it exists; add XMLString, XMLFile, and CharRef to it

refs [#5408](https://github.com/twisted/twisted/issues/5408)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun commented|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/numcharref-5408)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_7"></a>|@glyph set owner to @exarkun|
|-|-|

Looks fine; pretty trivial, with all the requirements hit; please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set status to `closed`|
|-|-|

(In [33418]) Merge numcharref-5408

Author: exarkun
Reviewer: glyph
Fixes: [#5408](https://github.com/twisted/twisted/issues/5408)

Add `twisted.web.template.CharRef`, for producing numeric character references
in documents generated using `twisted.web.template`.

This can be useful to generate pure-ASCII documents which nevertheless
contain non-ASCII characters.
